baseplate/crypto.py

```diff
@@ -92,89 +122,97 @@ class SignatureInfo(_SignatureInfo):
     pass
 
 
-class MessageSigner(object):
-    """Helper which signs messages and validates signatures given a secret.
+def _compute_digest(secret_value, header, message):
+    payload = header + message.encode("utf8")
+    digest = hmac.new(secret_value, payload, hashlib.sha256).digest()  # pylint: disable=no-member
+    return digest
 
-    .. testsetup::
 
-        import datetime
-        from baseplate.crypto import MessageSigner
+def make_signature(secret, message, max_age):
+    """Return a signature for the given message.
 
-    .. testcode::
+    To ensure that key rotation works automatically, always fetch the secret
+    token from the secret store immediately before use and do not cache / save
+    the token anywhere. The ``current`` version of the secret will be used to
+    sign the token.
 
-        message = "Hello, world!"
+    :param baseplate.secrets.VersionedSecret secret: The secret signing key
+        from the secret store.
+    :param str message: The message to sign.
+    :param datetime.timedelta max_age: The amount of time in the future
+        the signature will be valid for.
+    :return: An encoded signature.
 
-        signer = MessageSigner(b"supersecret")
-        signature = signer.make_signature(
-            message, max_age=datetime.timedelta(days=1))
+    """
+    version = 1
+    expiration = int(time.time() + max_age.total_seconds())
+    header = _HEADER_FORMAT.pack(version, expiration)
+    digest = _compute_digest(secret.current, header, message)
+    return base64.urlsafe_b64encode(header + digest)
+
+
+def validate_signature(secret, message, signature):
```

Comment:
Also thanks for linking that. It makes me feel better to see their implementation of multi-version on the read path is the same. I was a bit worried there'd be some timing attack concern I wasn't thinking about.